Yank/multistate/multistatesampler.py

```diff
@@ -886,26 +886,53 @@ def _restore_sampler_from_reporter(self, reporter):
         # data written just before a crash.
-        iteration = reporter.read_last_iteration()
-
-        # Retrieve other attributes.
         logger.debug("Reading storage file {}...".format(reporter.filepath))
-        thermodynamic_states, unsampled_states = reporter.read_thermodynamic_states()
-        sampler_states = reporter.read_sampler_states(iteration=iteration)
-        state_indices = reporter.read_replica_thermodynamic_states(iteration=iteration)
-        energy_thermodynamic_states, neighborhoods, energy_unsampled_states = reporter.read_energies(iteration=iteration)
-        n_accepted_matrix, n_proposed_matrix = reporter.read_mixing_statistics(iteration=iteration)
         metadata = reporter.read_dict('metadata')
+        thermodynamic_states, unsampled_states = reporter.read_thermodynamic_states()
 
-        # Search for last cached free energies only if online analysis is activated.
-        if self.online_analysis_interval is not None:
-            online_analysis_info = self._read_last_free_energy(reporter, iteration)
-            last_mbar_f_k, (_, last_err_free_energy) = online_analysis_info
-        else:
-            last_mbar_f_k, last_err_free_energy = None, None
+        def _read_options(check_iteration):
+            internal_sampler_states = reporter.read_sampler_states(iteration=check_iteration)
+            internal_state_indices = reporter.read_replica_thermodynamic_states(iteration=check_iteration)
+            internal_energy_thermodynamic_states, internal_neighborhoods, internal_energy_unsampled_states = \
+                reporter.read_energies(iteration=check_iteration)
+            internal_n_accepted_matrix, internal_n_proposed_matrix = \
+                reporter.read_mixing_statistics(iteration=check_iteration)
+
+            # Search for last cached free energies only if online analysis is activated.
+            if self.online_analysis_interval is not None:
+                online_analysis_info = self._read_last_free_energy(reporter, check_iteration)
+                internal_last_mbar_f_k, (_, internal_last_err_free_energy) = online_analysis_info
+            else:
+                internal_last_mbar_f_k, internal_last_err_free_energy = None, None
+            return (internal_sampler_states, internal_state_indices, internal_energy_thermodynamic_states,
+                    internal_neighborhoods, internal_energy_unsampled_states, internal_n_accepted_matrix,
+                    internal_n_proposed_matrix, internal_last_mbar_f_k, internal_last_err_free_energy)
+
+        # Keep trying to resume further and further back from the most recent checkpoint back
+        checkpoints = reporter.read_checkpoint_iterations()
+        checkpoint_reverse_iter = iter(checkpoints[::-1])
+        while True:
+            try:
+                checkpoint = next(checkpoint_reverse_iter)
+                output_data = _read_options(checkpoint)
```

Comment:
It may be worth testing if the netcdf file doesn't get messed up after the error. If it does, we might have to have this loop outside the whole `_restore_sampler_from_reporter()` method.

Reply:
What would we be checking for here. What is the exact metric of the netcdf file is "messed up?" And the sampler will have no idea about the netcdf file in theory, so we would have to make a `Reporter` method which does like a integrity check either as a whole or at an iteration.

Reply:
Sorry, by testing I didn't mean writing an actual test. I'm aware this is very hard to reproduce. I meant we should be on the lookout since after the first HDF5 error (or similar), the netcdf dataset object could be left in an inconsistent state and we may have to close the dataset and open a new one instead of reading previous iterations from the same dataset object.

> And the sampler will have no idea about the netcdf file in theory

I agree. I was only proposing as a possible solution to the problem above (and only in the case the problem was real) to move the `while True` loop from `_restore_sampler_from_reporter()` to `from_storage()` to close/open the dataset at every loop iteration.

Nothing that needs to be done if we don't incur any problems with this implementation though!